namer/namer.cc

```diff
@@ -15,392 +18,730 @@
- * table.
- */
-class NameInserter {
-    friend class Namer;
+namespace {
+class FoundNames;
+
+struct FoundClassRef;
+struct FoundClass;
+struct FoundFieldOrVariable;
+struct FoundTypeMember;
+struct FoundMethod;
+
+enum class NameKind : u1 {
+    None = 0,
+    Root = 1,
+    Class = 2,
+    ClassRef = 3,
+    Method = 4,
+    FieldOrVariable = 5,
+    TypeMember = 6,
+    Symbol = 7,
+};
 
-    core::SymbolRef squashNames(core::MutableContext ctx, core::SymbolRef owner, unique_ptr<ast::Expression> &node) {
-        auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(node.get());
-        if (constLit == nullptr) {
-            if (auto *id = ast::cast_tree<ast::ConstantLit>(node.get())) {
-                return id->symbol.data(ctx)->dealias(ctx);
-            }
-            if (auto *uid = ast::cast_tree<ast::UnresolvedIdent>(node.get())) {
-                if (uid->kind != ast::UnresolvedIdent::Kind::Class || uid->name != core::Names::singleton()) {
-                    if (auto e = ctx.beginError(node->loc, core::errors::Namer::DynamicConstant)) {
-                        e.setHeader("Unsupported constant scope");
-                    }
-                }
-                // emitted via `class << self` blocks
-            } else if (ast::isa_tree<ast::EmptyTree>(node.get())) {
-                // ::Foo
-            } else if (node->isSelfReference()) {
-                // self::Foo
-            } else {
-                if (auto e = ctx.beginError(node->loc, core::errors::Namer::DynamicConstant)) {
-                    e.setHeader("Dynamic constant references are unsupported");
-                }
-            }
-            node = ast::MK::EmptyTree();
-            return owner;
-        }
+class FoundNameRef final {
+    u4 _id;
 
-        auto newOwner = squashNames(ctx, owner, constLit->scope);
-        core::SymbolRef existing = newOwner.data(ctx)->findMember(ctx, constLit->cnst);
-        if (!existing.exists()) {
-            if (!newOwner.data(ctx)->isClassOrModule()) {
-                if (auto e = ctx.beginError(node->loc, core::errors::Namer::InvalidClassOwner)) {
-                    auto constLitName = constLit->cnst.data(ctx)->show(ctx);
-                    auto newOwnerName = newOwner.data(ctx)->show(ctx);
-                    e.setHeader("Can't nest `{}` under `{}` because `{}` is not a class or module", constLitName,
-                                newOwnerName, newOwnerName);
-                    e.addErrorLine(newOwner.data(ctx)->loc(), "`{}` defined here", newOwnerName);
-                }
-                node = ast::MK::EmptyTree();
-                return owner;
-            }
-            existing = ctx.state.enterClassSymbol(core::Loc(ctx.file, constLit->loc), newOwner, constLit->cnst);
-            existing.data(ctx)->singletonClass(ctx); // force singleton class into existance
-        }
+    // Store NameKind in upper 8 bits of idx.
+    static u4 mask(NameKind kind, u4 idx) {
```

Comment:
Do you get much with this binpacking?
It's a timebomb so let's see how valuable it is and otherwise remove it.
I don't expect it to earn you much

Reply:
+1

Reply:
I removed it. I don't think it's that dangerous (will we ever see more than 2^24 symbols in one file?) but I also doubt it buys us that much.